### PR TITLE
4531 cloudmc source

### LINKED ISSUE #4531: cloudmc settings source editor
Add an editor for the openmc settings.source to the visualization page. Allow adding a group different source types in one container (Point, Box, SphericalIndependent, etc).

https://docs.openmc.org/en/stable/usersguide/settings.html#external-source-distributions

### PR DESCRIPTION
**Pull request by moellep**
- fix #4531 implement a hierarchical source editor for various geometries/distributions
- use pymoab library for processing dagmc files

### CODE REVIEW

**Comment by moellep** [comment]
@mkeilman I've emailed you the three example sirepo-data.json files (fuel-pin, reactor, and utah_teapot).

File: sirepo/package_data/static/json/cloudmc-schema.json [hunk 1]
```diff
@@ -121,34 +187,155 @@
...
         },
         "tally": {
             "aspect": ["Aspect to plot", "TallyAspect", "mean"],
             "aspects": ["Aspects", "TallyAspects", ["mean"]],
-            "name": ["Name", "String"],
-            "meshCellCount": ["Mesh cell count", "IntStringArray", "10, 10, 10", "", "X, Y, Z"],
-            "meshLowerLeft": ["Mesh lower left", "Float3", "-5.0, -5.0, -5.0", "", "X, Y, Z"],
-            "meshUpperRight": ["Mesh upper right", "Float3", "5.0, 5.0, 5.0", "", "X, Y, Z"],
+            "name": ["Name", "String", "t1"],
+            "meshCellCount": ["Mesh cell count", "Point3D", [10, 10, 10]],
```
Comment (mkeilman) [inline sirepo/package_data/static/json/cloudmc-schema.json:265]: See earlier comment about validating ints and value min/max

**Review by mkeilman — COMMENTED** [review]
Nothing substantial. Style stuff, plus a comment on "Point3D" which should not hold this up

File: sirepo/package_data/static/js/cloudmc.js [hunk 2]
```diff
@@ -835,6 +874,241 @@ SIREPO.app.directive('componentName', function(appState, requestSender) {
...
+        template: `
+            <div data-ng-repeat="v in model[field] track by $index"
+              style="display: inline-block; width: 7em; margin-right: 5px;" >
+              <input class="form-control" data-string-to-number="Float"
+                data-ng-model="model[field][$index]"
+                style="text-align: right" required />
+            </div>
+        `,
+    };
+});
```
[the remark this answers is not in the document]
Comment (moellep) [reply sirepo/package_data/static/js/cloudmc.js:964]: Yes, a string array of numbers isn't a very good format. They UI layout for Point3D is a bit better than numberList, and it edits all the values as float in place so it doesn't require any parsing. It doesn't handle integers or ranges. I think it is good enough for this app for now, and I'll add a float-to-int conversion on the server just to be safe:
```
v.tallyMeshCellCount = _generate_array([int(v) for v in data.models.tally.meshCellCount])
```

File: sirepo/template/cloudmc.py [hunk 3]
```diff
@@ -87,25 +87,174 @@ def write_parameters(data, run_dir, is_parallel):
...
+    args = []
+    if angle._type == "isotropic":
+        pass
+    elif angle._type == "monodirectional":
+        args.append(_generate_array(angle.reference_uvw))
+    elif angle._type == "polarAzimuthal":
+        args += [_generate_angleribution(angle[v] for v in ["mu", "phi"])]
+        args.append(_generate_array(angle.reference_uvw))
+    else:
+        raise AssertionError("unknown angle type: {}".format(angle._type))
```
[the remark this answers is not in the document]
Comment (moellep) [reply sirepo/template/cloudmc.py:102]: I think it is OK to raise an assertion on the final "else". That way you don't need to update the assert clause if you add a new "if" handler - the assert fails only when it fails to find a handler.

File: sirepo/package_data/static/js/cloudmc.js [hunk 4]
```diff
@@ -835,6 +874,241 @@ SIREPO.app.directive('componentName', function(appState, requestSender) {
...
+                    if (m._index != model._index) {
+                        m._index = c.length;
+                        c.push(m);
+                    }
+                }
+                $scope.model[$scope.field] = c;
+            };
+
+            $scope.$on('modelChanged', function(event, name) {
+                if (name == childModel) {
```
[the remark this answers is not in the document]
Comment (moellep) [reply sirepo/package_data/static/js/cloudmc.js:1101]: In our case, childModel can never equal the outer $scope.modelName. It is a tree structure, but no model can have children of its own type.

**Pull request merged**